src/bootstrap/step.rs

```diff
@@ -1,1820 +0,0 @@
-
-        rules.into_iter().flat_map(|(rule, _)| {
-            let hosts = if rule.only_host_build || rule.only_build {
-                self.build.build_slice()
-            } else {
-                &self.build.hosts
-            };
-            // Determine the actual targets participating in this rule.
-            // NOTE: We should keep the full projection from build triple to
-            // the hosts for the dist steps, now that the hosts array above is
-            // truncated to avoid duplication of work in that case. Therefore
-            // the original non-shadowed hosts array is used below.
-            let arr = if rule.host {
-                // If --target was specified but --host wasn't specified,
-                // don't run any host-only tests.
-                if self.build.flags.host.len() > 0 {
-                    &self.build.hosts
-                } else if self.build.flags.target.len() > 0 {
-                    &[]
-                } else if rule.only_build {
-                    self.build.build_slice()
-                } else {
-                    &self.build.hosts
-                }
-            } else {
-                &self.build.targets
-            };
-
-            hosts.iter().flat_map(move |host| {
-                arr.iter().map(move |target| {
-                    self.sbuild.name(rule.name).target(target).host(host)
-                })
-            })
-        }).collect()
-    }
-
-    /// Execute all top-level targets indicated by `steps`.
-    ///
-    /// This will take the list returned by `plan` and then execute each step
-    /// along with all required dependencies as it goes up the chain.
-    fn run(&self, steps: &[Step<'a>]) {
-        self.build.verbose("bootstrap top targets:");
-        for step in steps.iter() {
-            self.build.verbose(&format!("\t{:?}", step));
-        }
-
-        // Using `steps` as the top-level targets, make a topological ordering
-        // of what we need to do.
-        let order = self.expand(steps);
-
-        // Print out what we're doing for debugging
-        self.build.verbose("bootstrap build plan:");
-        for step in order.iter() {
-            self.build.verbose(&format!("\t{:?}", step));
-        }
-
-        // And finally, iterate over everything and execute it.
-        for step in order.iter() {
-            if self.build.flags.keep_stage.map_or(false, |s| step.stage <= s) {
-                self.build.verbose(&format!("keeping step {:?}", step));
-                continue;
-            }
-            self.build.verbose(&format!("executing step {:?}", step));
-            (self.rules[step.name].run)(step);
-        }
-
-        // Check for postponed failures from `test --no-fail-fast`.
-        let failures = self.build.delayed_failures.get();
-        if failures > 0 {
-            println!("\n{} command(s) did not execute successfully.\n", failures);
-            process::exit(1);
```

Comment:
This is the code that @infinity0 is talking about.  The `delayed_failures` counter still exists, and is incremented, but AFAICS nothing checks it anymore.

Reply:
Looks like this just needs to go at the end of `Build::build`.  I'll try that and send a PR.

Reply:
Yeah, it's quite likely that we're just not checking it anymore. r? me on the PR please!

Reply:
@infinity0 has a change to also summarize the failed commands, so I think they'll PR it.